### ISSUE TITLE
Improper DEFAULT_MAX_NURSERY?

### ISSUE THREAD

**Issue by qinsoon** [mmtk/mmtk-core#723]
Our current `DEFAULT_MAX_NURSERY` is set to 1G on 64bits machine.
https://github.com/mmtk/mmtk-core/blob/9edb4539826bf733059c9c33106922ad972785bf/src/util/options.rs#L83

According to `collection_required()` in `Gen`, nursery is not considered as full (to trigger a nursery GC) unless its reserved pages reach the max nursery size (1G).
https://github.com/mmtk/mmtk-core/blob/9edb4539826bf733059c9c33106922ad972785bf/src/plan/generational/global.rs#L114-L115

If the heap size is smaller than 1G and we do not set nursery size explicitly, we always trigger full heap GCs, and never trigger a nursery GC.

**Comment by k-sareen** [comment]
> If the heap size is smaller than 1G and we do not set nursery size explicitly, we always trigger full heap GCs, and never trigger a nursery GC.

That's not true, right? I don't remember noticing this while I fixed the nursery size options. The function which actually controls a full-heap GC is `require_full_heap_collection` [1], not `collection_required`.

\[1]: https://github.com/mmtk/mmtk-core/blob/9edb4539826bf733059c9c33106922ad972785bf/src/plan/generational/global.rs#L147-L193

**Comment by qinsoon** [comment]
If `nursery.reserved_pages` is always smaller than `nursery_max_size`, then we will not trigger a GC because of nursery being full. In that case, we will trigger a GC if `plan.reserved_pages` is larger `plan.total_pages`, and that will be a full heap GC.

**Comment by k-sareen** [comment]
So in other words, only highly allocating/generational workloads work as intended... Hmm. From memory we discussed that a large nursery limit was fine since it's the "unbounded" nursery. Surely one can argue that nursery size is a tuning parameter specific to each workload and in order to get best performance, they may have to tune it?

**Comment by qinsoon** [comment]
> So in other words, only highly allocating/generational workloads work as intended... 

It is not about work load. It is max nursery size and heap size. If max nursery size is larger than heap size, we will not have nursery GCs. For example, the default max nursery value (1G) is very likely to be larger than normal 2x-6x heap sizes we use for most dacapo benchmarks. If that happens, we will not see nursery GCs.

I believe we have the same issue for Java MMTk. But their default max nursery is 32M, which is more likely to be smaller than usual heap sizes we use for Dacapo.

> Hmm. From memory we discussed that a large nursery limit was fine since it's the "unbounded" nursery. Surely one can argue that nursery size is a tuning parameter specific to each workload and in order to get best performance, they may have to tune it?

Right. You would need to set nursery size along with heap size, and make sure they match each other.

**Comment by qinsoon** [comment]
For example, the following always triggers full heap GCs. The default max nursery is 1G, and the heap size is set to 80M. The heap is full before the nursery is full -- so every GC is full heap.
```
MMTK_PLAN=GenImmix build/linux-x86_64-normal-server-release/jdk/bin/java -XX:MetaspaceSize=500M -XX:+DisableExplicitGC -XX:-TieredCompilation -XX:+UseThirdPartyHeap -Xms80M -Xmx80M -cp /usr/share/benchmarks/dacapo/dacapo-2006-10-MR2.jar Harness fop
```
```console
[2023-01-02T22:40:11Z INFO  mmtk::memory_manager] Initialized MMTk with GenImmix
===== DaCapo fop starting =====
[2023-01-02T22:40:12Z INFO  mmtk::util::heap::gc_trigger] [POLL] nursery: Triggering collection
[2023-01-02T22:40:12Z INFO  mmtk::plan::generational::global] Full heap GC
[2023-01-02T22:40:12Z INFO  mmtk::scheduler::gc_work] End of GC
[2023-01-02T22:40:13Z INFO  mmtk::util::heap::gc_trigger] [POLL] nursery: Triggering collection
[2023-01-02T22:40:13Z INFO  mmtk::plan::generational::global] Full heap GC
[2023-01-02T22:40:13Z INFO  mmtk::scheduler::gc_work] End of GC
===== DaCapo fop PASSED in 1348 msec =====
```

If you set `MMTK_NURSERY=Bounded:30000000` (30M) to the above command, we will see nursery GC.
```
MMTK_NURSERY=Bounded:30000000 MMTK_PLAN=GenImmix build/linux-x86_64-normal-server-release/jdk/bin/java -XX:MetaspaceSize=500M -XX:+DisableExplicitGC -XX:-TieredCompilation -XX:+UseThirdPartyHeap -Xms80M -Xmx80M -cp /usr/share/benchmarks/dacapo/dacapo-2006-10-MR2.jar Harness fop
```
```console
[2023-01-02T22:26:19Z INFO  mmtk::memory_manager] Initialized MMTk with GenImmix
===== DaCapo fop starting =====
[2023-01-02T22:26:20Z INFO  mmtk::util::heap::gc_trigger] [POLL] nursery: Triggering collection
[2023-01-02T22:26:20Z INFO  mmtk::plan::generational::global] Nursery GC
[2023-01-02T22:26:20Z INFO  mmtk::scheduler::gc_work] End of GC
[2023-01-02T22:26:20Z INFO  mmtk::util::heap::gc_trigger] [POLL] nursery: Triggering collection
[2023-01-02T22:26:20Z INFO  mmtk::plan::generational::global] Nursery GC
[2023-01-02T22:26:20Z INFO  mmtk::scheduler::gc_work] End of GC
[2023-01-02T22:26:20Z INFO  mmtk::util::heap::gc_trigger] [POLL] nursery: Triggering collection
[2023-01-02T22:26:20Z INFO  mmtk::plan::generational::global] Nursery GC
[2023-01-02T22:26:20Z INFO  mmtk::scheduler::gc_work] End of GC
===== DaCapo fop PASSED in 1359 msec =====
```

**Comment by k-sareen** [comment]
This run is from a while ago but is from my PR comment here (https://github.com/mmtk/mmtk-core/pull/630#issuecomment-1197562684), but it seems like the `majorGC != GC` [1]. Maybe the counter is broken then...? I guess we need to remove the `is_nursery_full` check and replace it with something else.

[1]: https://squirrel.anu.edu.au/plotty/kunals/bench-test/#0|2022-07-27-nursery-pr-performance-boar-2022-07-27-Wed-153152&benchmark^build^buildstring^invocation^iteration^mmtk_gc&GC^majorGC^time^time.other^time.stw&|10&iteration^1^4&mmtk_gc^1^GenImmix|70&build^nursery-pr

**Comment by k-sareen** [comment]
Yes this is reproducible on current mmtk-core [`master`](https://github.com/mmtk/mmtk-core/tree/06237f15039d08b9faa710ce4e32a422fd98996e) + mmtk-openjdk [`master`](https://github.com/mmtk/mmtk-openjdk/tree/b01e1cb235027b5441e46ef4e3c1969875ad2bff).

For `GenImmix` + `fop`:
```
============================ MMTk Statistics Totals ============================
GC      time.other      time.stw        majorGC total-work.time.min     total-work.time.max     total-work.time.total   total-work.count
16      850.32  71.35   16      inf     -inf    0.000   0
Total time: 921.67 ms
------------------------------ End MMTk Statistics -----------------------------
===== DaCapo evaluation-git-6e411f33 fop PASSED in 922 msec =====
```

But not reproducible on mmtk-core [`8afe96d`](https://github.com/mmtk/mmtk-core/tree/8afe96d7931e26b957e52bcd7f7d271510641a16) and mmtk-openjdk [`688791e`](https://github.com/mmtk/mmtk-openjdk/commit/688791ec3eeb394fbc07d0533844f65ec8df9e13), which actually introduced the bounded nursery.

For `GenImmix` + `fop`:
```
============================ MMTk Statistics Totals ============================
GC      time.other      time.stw        majorGC total-work.time.max     total-work.time.min     total-work.time.total   total-work.count
48      940.76  151.32  2       -inf    inf     0.000   0
Total time: 1092.08 ms
------------------------------ End MMTk Statistics -----------------------------
===== DaCapo evaluation-git-6e411f33 fop PASSED in 1092 msec =====
```

The static cost of a GC unfortunately outweighs the gains from nursery GC as discussed here #594 and previous meetings.

**Comment by qinsoon** [comment]
Thanks @k-sareen . I figured out why it could trigger nursery GCs in [8afe96d](https://github.com/mmtk/mmtk-core/tree/8afe96d7931e26b957e52bcd7f7d271510641a16), but not in the current [master](https://github.com/mmtk/mmtk-core/tree/06237f15039d08b9faa710ce4e32a422fd98996e). The problem I talked about in this issue also exists in [8afe96d](https://github.com/mmtk/mmtk-core/tree/8afe96d7931e26b957e52bcd7f7d271510641a16): GC is triggered because of heap being full (not nursery being full). The difference is after the GC is triggered, `master` always treats the GC as full heap GC (which is obvious as the heap was full), but `8afe96d` may treat it as nursery GC.

The difference is from another change in https://github.com/mmtk/mmtk-core/commit/2300ce1d4a7c99053c5bb7055b377a71c9241e85. L72 (`pr.clear_request(pages_reserved)`) was before `block_for_gc()` in `8afe96d`, but it is moved to after `block_for_gc()` in 2300ce1d4a7c99053c5bb7055b377a71c9241e85 (this change was for a different reason). So in `8afe96d`, there are a few pending allocation pages that are counted when we decide to trigger GC, but they are not counted when we decide nursery/full heap GC. 
https://github.com/mmtk/mmtk-core/blob/06237f15039d08b9faa710ce4e32a422fd98996e/src/policy/space.rs#L69-L72

This is what happens for `8afe96d`:
```
MMTK_PLAN=GenImmix build/linux-x86_64-normal-server-release/jdk/bin/java -XX:MetaspaceSize=500M -XX:+DisableExplicitGC -XX:-TieredCompilation -XX:+UseThirdPartyHeap -Xms80M -Xmx80M -cp /usr/share/benchmarks/dacapo/dacapo-2006-10-MR2.jar Harness fop
```
```console
[2023-01-04T04:52:52Z TRACE mmtk::plan::generational::global] nursery = 9864, max nursery =268435456
// Nursery is not full
[2023-01-04T04:52:52Z DEBUG mmtk::plan::global] self.get_reserved_pages()=20483, self.get_total_pages()=20480
// But heap is full - Do GC
[2023-01-04T04:52:52Z INFO  mmtk::plan::global]   [POLL] nursery: Triggering collection
// Clear current pending allocation pages (reserved pages is now 20467 down from 20483)
[2023-01-04T04:52:52Z TRACE mmtk::plan::generational::global] full heap? false (total 20480, reserve 20467)
// So heap is not full - Do nursery GC
[2023-01-04T04:52:52Z INFO  mmtk::plan::generational::global] Nursery GC
[2023-01-04T04:52:52Z INFO  mmtk::scheduler::gc_work] End of GC
```
Though this triggers a nursery GC, I don't feel this is intended. I would say it was a bug that makes it possible to trigger nursery GCs in `8afe96d`, the problem we discussed in this issue still exists in `8afe96d`.

**Comment by k-sareen** [comment]
Going by the GC Handbook, what we want is an Appel-style nursery (Chapter 9, pg. 121). A heap-full collection should be always be nursery GC _unless_ the nursery space is less than some minimum threshold (i.e. the `should_next_gc_be_full_heap()` function). So really, it's our way of triggering nursery GCs that is incorrect now.

**Comment by k-sareen** [comment]
I think the terminology of "heap-full" and "full-heap" is perhaps confusing us. Heap-full is just when we are out of space to allocate. I don't see why a full-heap GC is required for that case. A full-heap GC should only happen if the nursery is too small to allocate further. (Maybe we should rename "heap-full" to something else?)
